### ISSUE TITLE
Component does not load js file, when "-" or "/" in register name

### ISSUE THREAD

**Issue by M4p4** [EmilStenstrom/django-components#262]
Given is following code:
```python
@component.register("foo-card")
class Foo(component.Component):
      template_name = "foo-card/foo-card.html"

      def get_context_data(self, url):
          return {
              "url": url,
          }

      class Media:
         js = "foo-card/foo-card.js"
```     

The component is registered with the name "foo-card". When this component is used the js file `foo-card.js` does not load!

When i change the register name of the component to:

```python
@component.register("foocard")
```
the component and the js file `foo-card.js` load!

I think the js should also load when there is a - or / in the component name.

**Comment by EmilStenstrom** [comment]
@lemontheme I vaguely remember something about component names being valid identifiers?

@M4p4 Are you open to supplying a PR that fixes this?

**Comment by lemontheme** [comment]
Ah, beat me to it @EmilStenstrom!

This is a bug. I'll explain why below. Fortunately, this one's easy to set right. I can easily patch this before the end of the day (CEST).

What's happened here is that, around the time we introduced fill tag aliases, we experimented with treating component names, fill names, and slot names as identifiers rather than as variables that need resolving. To illustrate the difference,
the names of Django's builtin 'block' tags are treated identifiers, e.g. `title` below, which has a constant value = 'title', regardless of the rendering context.

```django
{% block title %}{{ section.title }}{% endblock %}
```

By contrast, the argument of a builtin  'if' tag is treated as a variable, i.e. it is resolved against the rendering context. For example, the value of `athlete_list` below depends on the value associated with the corresponding key in the redering context.

```django
{% if athlete_list %} ...
```

In this setting, if the argument is quotes it is treated as a literal string. E.g. `"athlete_list"` resolves to `"athlete list"`.

```django
{% if "athlete_list" %} ...
```

In the end, we determined it made most sense to treat component names, fill names, and slot names as variables rather than as identifiers. Because variables only allow `_` as a separator, I (mistakenly) figured it would mean less confusion in the long run if we extended that naming constraint (i.e. that all names must be valid python identifiers) to component names – i.e. such that a component name like `foo-card` would be rejected, but `foo_card` would be permitted. Clever me forgot to add a check to the register() function, so that's why this didn't raise an error when registering the component.

In the case in OP, the JS doesn't load because the component doesn't load, and looking at the middleware implementation, I believe I'm seeing that media (JS, CSS) is added if and only if the component is loaded.  (I'm not too familiar with the middleware part of the codebase, but I think that's what's happening.)

Anyway, the strict naming constraint on component names is no longer appropriate. Relaxing it should fix this issue. I'll also see if I can't add a regression test for the case in OP.

**Comment by M4p4** [comment]
@lemontheme  Thanks for the detailed explanation, it would be cool if you can fix it, since I go tomorrow into holidays. But next time, if I find a bug, I will accept the challenge. :D

**Comment by timothyis** [comment]
Just for what it's worth, I think it'd be valuable to this repo to include an issue template asking for the version that the issue creator is using @EmilStenstrom @lemontheme. 

[This comment](https://github.com/EmilStenstrom/django-components/issues/262#issuecomment-1505019381) is assuming that we're on a version that has changes (I think).

Just to be clear, we're currently using 0.25, where this issue is present. We are updating soon.

**Comment by lemontheme** [comment]
Excellent suggestion. I was so wrapped up in finding a fix it didn't even occur to me to ask about the version. Will go back to the v0.25 commit to check if there's a pattern.

**Comment by lemontheme** [comment]
@M4p4 – Any chance you could give me an excerpt of the template where you're including the 'foo-card' component?

**Comment by M4p4** [comment]
```html
<section class="bg-primary-25">
        <div class="container mx-auto px-8 py-6 lg:pt-10 lg:pb-20">
		<div class="grid grid-cols-12 gap-y-4 md:gap-x-8 min-h-0 max-h-fit grow self-start mb-8">
			{% for url in url_list %}
				<div class="col-span-12 md:col-span-6 xl:col-span-4 min-h-0">
					{% component "foo-card" url=url only %}
				</div>
			{% endfor %}
		</div>
	</div>
</section> 
```

**Comment by lemontheme** [comment]
Thanks! Nothing out of the ordinary there.

I'm still trying to figure out what's going wrong. Strangely enough, with the latest versions quoted component names, e.g. `"foo-card"`, work just fine. I'm now working back through various versions to see if the behavior there is different.

**Comment by timothyis** [comment]
> @M4p4, mind giving me the version you're on?

I work with @M4p4, we're on 0.25.

Some additional context is that we are not using automatic component registering. We define our components in our settings because we had a component called `Select` and it was then breaking something with the same name.

**Comment by lemontheme** [comment]
Alright, then I've got two tiny more questions for you, since I'm kind of scratching my head at the moment.

1. Wrt the original post, am I understanding correctly that the component template itself renders, regardless of the name? It's just the JS media that doesn't load?
2. Are you using any of the django-component middlewares? (They're still rather under-documented.)

**Comment by lemontheme** [comment]
Ah-ha, we have a culprit! This issue is entirely unrelated to what I assumed it would be. It's `ComponentDependencyMiddleware` that we need to take a closer look at.  (I'm new to this one. Took me a while to figure I need to set `render_dependencies=True` _in addition_ to including the middleware itself in order for it to even run. Bit strange, that.) Anyway, now I can repro the error, we're one step closer to figuring out the actual cause 🕵️‍♀️

**Comment by lemontheme** [comment]
Got it. The [regexes in middleware.py](https://github.com/EmilStenstrom/django-components/blob/master/django_components/middleware.py#LL12C1-L17C2) are too strict. Adding a hyphen and `/` to it seems to make the issue go away.

Will revisit this tomorrow. Once I'm satisified that making the regexes more general doesn't cause problems elsewhere, I'll open a PR.

**Comment by EmilStenstrom** [comment]
@lemontheme Yeah, the ComponentDependencyMiddleware is not in a great state documentation and usability wise. There's an issue open for that here: https://github.com/EmilStenstrom/django-components/issues/71

**Comment by EmilStenstrom** [comment]
I've just [released 0.28.2](ComponentDependencyMiddleware) which adds support for dash and slash in component names when using the ComponentDependencyMiddleware. Hope this removes a small confusion when trying out django-components.

**Comment by imankulov** [comment]
Hey folks! 👋

I was working on the same codebase as @M4p4 and @timothyis and came across the same issue. It seems that it hasn't been properly fixed initially, as only one out of the two regexes has been updated.

I have created a PR #320 which, hopefully, fixes it this time.
